examples/shors.rs

```diff
@@ -0,0 +1,289 @@
+use clap::Parser;
+
+use colored::Colorize;
+use log::debug;
+use num::traits::Pow;
+use quaru::math::{limit_denominator, modpow};
+use quaru::{operation, register::Register};
+use rand::Rng;
+use stopwatch::Stopwatch;
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Number to factorize
+    #[arg(short, long, default_value_t = 15)]
+    number: u32,
+
+    /// Number of times to run the algorithm
+    #[arg(short, long, default_value_t = 1)]
+    n_times: u32,
+}
+
+fn main() {
+    let args = Args::parse();
+    let number = args.number;
+    let n_times = args.n_times;
+
+    let mut runtimes = Vec::<i64>::new();
+    for _ in 0..n_times {
+        let sw = Stopwatch::start_new();
+        println!("Running for N = {}", number);
+
+        // Find a factor with shor's algorithm
+        let d1 = shors(number);
+        let d2 = number / d1;
+        println!(
+            "The factors of {} are {} and {}",
+            number.to_string().green(),
+            d1.to_string().blue(),
+            d2.to_string().yellow()
+        );
+        let t = sw.elapsed_ms();
+        runtimes.push(t);
+        println!("Time elapsed: {} ms", t.to_string().blue());
+        println!("------------------------------------")
+    }
+
+    let avg = runtimes.iter().sum::<i64>() / runtimes.len() as i64;
+    let min = runtimes.iter().min().unwrap();
+    println!(
+        "Total time: {} ms after {} runs",
+        runtimes.iter().sum::<i64>().to_string().red(),
+        n_times.to_string().red()
+    );
+    println!("Avarage time: {} ms", avg.to_string().bright_purple());
+    println!("Best time: {} ms", min.to_string().green());
+}
+
+/// Create a gate that multiplies its input by a^2^i mod N.
+fn u_gate(targets: Vec<usize>, modulus: u32, a: u32, i: usize) -> operation::Operation {
+    // Calculate a^2^i mod N
+    let a_pow_mod: usize = modpow(a, 1 << i, modulus) as usize;
+
+    debug!("a = {}, i = {}, mod = {}", a, i, modulus);
+    debug!("a^2^i % mod = {}", a_pow_mod);
+
+    // Create the function for the controlled u gate
+    let func = |x: usize| -> usize { (x * a_pow_mod) % modulus as usize };
+    // Create the gate
+    let u_gate = operation::to_quantum_gate(&func, targets.clone());
+
+    u_gate
+}
+
+/// Shor's algorithm
+/// This algorithm finds a factor of the number N.
+fn shors(number: u32) -> u32 {
+    // Shor's algorithm doesn't work for even numbers
+    if number % 2 == 0 {
+        return 2;
+    }
+
+    // Shor's algorithm doesn't work for prime powers
+    // Testing up to k=log2(N) is enough because if k > log2(N), 2^k > N.
+    // (Since we already tested for even numbers, testing up to log3(N) would actually be enough)
+    for k in 2..number.ilog2() + 1 {
```

Comment:
I get an error here so that my code does not compile: "no method `ilog2` on type `u32`"

Anyone else?